bittide/src/Clash/Cores/UART/Extra.hs

```diff
@@ -0,0 +1,66 @@
+-- SPDX-FileCopyrightText: 2023 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# OPTIONS_GHC -fplugin Protocols.Plugin #-}
+module Clash.Cores.UART.Extra
+  ( module Clash.Cores.UART.Extra
+  , System.IO.stdin
+  , System.IO.stdout
+  ) where
+
+import Clash.Prelude
+
+import Clash.Cores.UART
+import Data.Char
+import Data.Maybe
+import System.IO
+import Protocols
+import Protocols.Df hiding (catMaybes, sample, pure)
+import Protocols.Internal
+import GHC.IO
+
+import Bittide.Wishbone
+
+import qualified Protocols.Df as Df
+
+-- | A simulation function for circuits that expose a UART connection.
+-- This function reads from the provided input handle and feeds that to the UART circuit.
+-- Incoming uart data is written the the output handle.
+uartIO ::
+  forall dom baud .
+  (KnownDomain dom, ValidBaud dom baud) =>
+  -- | A handle for the input data stream, use `stdin` for terminal input.
+  Handle ->
+  -- | A handle for the output data stream, use `stdout` for terminal output.
+  Handle ->
+  -- | The baud rate for the UART communication.
+  SNat baud ->
+  -- | The circuit to be simulated.
+  Circuit (CSignal dom Bit) (CSignal dom Bit) ->
+  -- | The IO action that performs simulation.
+  IO ()
```

Comment:
This will take complete "control" of simulating the circuit, there might be cases where we want to involve a UART but run simulation in another form (for testing or debugging).

Could it make sense to have a more "general purpose" version of this and then have one version that runs everything as one `IO ()` action?